README.md
# commctl
Commissaire command line interface for [commissaire](https://github.com/projectatomic/commissaire)

[![Documentation](https://readthedocs.org/projects/commissaire/badge/?version=latest)](http://commissaire.readthedocs.org/) [![Build Status](https://travis-ci.org/projectatomic/commctl.svg)](https://travis-ci.org/projectatomic/commctl)


CHANGELOG.md
# commctl v0.0.6
```
* 4c9bb6a: travis: Force lib upgrade on tests
* 992c992: doc: Update copyright year
* a5f18c3: test: Use six.b for _content in mocks
* c3419d1: test: Add network create to cli tests
* 9d5d0db: test: Add client script function tests
* 387566d: Post-release version bump.
```

# commctl v0.0.5
```
* 81dcfbd: cli: Rework --options for "container_manager create".
* 1b77f09: cli: Enable --type when creating container manager.
* 687bddc: Post-release version bump.
```

# commctl v0.0.4
```
* 2602c30: cli: Add host join/unjoin commands.
* bde5a30: cli: Add user-data command
* 97b4acd: setup: Add data dir for future user-data command
* 1a6a65e: helpers: passhash moved
* 2a82a25: cli: Example updated to point to deploy
* 4efabce: Use "six" for easier Python 2/3 compatibility.
* 3ec58a2: Support Python 3.
* 85bf438: Post-release version bump.
```

# commctl v0.0.3
```
* 6b31ee2: Post-release version bump.
```

# commctl v0.0.2
```
* 185f47c: Add CHANGELOG.md for 0.0.2 release.
* 1e2d728: Fix container manager routes.
* b54ba92: container: Added Dockerfile
* a0b7ab5: Added new command container_manager.
```


test/__init__.py
# Copyright (C) 2016-2017  Red Hat, Inc
#
# Copyright (C) 2016-2017  Red Hat, Inc
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Library General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Library General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301  USA
"""
Unittests.
"""

import os
import unittest


def get_fixture_file_path(filename):
    """
    Attempts to return the path to a fixture file.

    :param filename: The name of the file to look for.
    :type filename: str
    :returns: Full path to the file
    :rtype: str
    :raises: Exception
    """
    for x in ('.', '..'):
        try:
            a_path = os.path.sep.join((x, filename))
            os.stat(a_path)
            return os.path.realpath(a_path)
        except:
            pass
    raise Exception(
        'Can not find path for config: {0}'.format(filename))


class TestCase(unittest.TestCase):
    """
    Parent class for all unittests.
    """
    pass


test/test_cli.py
# Copyright (C) 2016-2017  Red Hat, Inc
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Library General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Library General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301  USA
"""
Test cases for the commctl.client_script script.

.. note:: The use of \ in the with/mock blocks is required for readability.
          Unfortunately, ()'s for multiline statements do not work with
          the with structure in 2.7.
"""
import argparse
import json
import socket

import six

if six.PY2:
    import mock
else:
    from unittest import mock

from . import TestCase, get_fixture_file_path
from commctl import cli


class TestCliHostAddress(TestCase):
    """
    Tests for cli.host_address function.
    """

    def test_host_address(self):
        """
        Verify host_address resolves to a host name.
        """
        with mock.patch('socket.gethostbyname') as _ghn:
            _ghn.return_value = 'example.org'
            result = cli.host_address('192.168.0.1')

            self.assertEquals(
                result,
                _ghn.return_value)

    def test_host_address_unable_to_resolve(self):
        """
        Verify host_address raises on failed resolves.
        """
        with mock.patch('socket.gethostbyname') as _ghn:
            _ghn.side_effect = socket.gaierror('test')
            self.assertRaises(
                argparse.ArgumentTypeError,
                cli.host_address,
                '192.168.0.1')


class TestCliAssembleOptions(TestCase):
    """
    Tests for cli.assemble_options function.
    """

    def test_assemble_options_with_equals(self):
        """
        Verify assemble_options works with ['a=b',...].
        """
        data = ['a=b', 'c=d']
        expected = {'a': 'b', 'c': 'd'}
        self.assertEquals(
            cli.assemble_options(data),
            expected)

    def test_assemble_options_with_list_of_json(self):
        """
        Verify assemble_options works with list of json.
        """
        expected = {'a': 'b', 'c': 'd'}
        self.assertEquals(
            cli.assemble_options([json.dumps(expected)]),
            expected)

    def test_assemble_options_with_bad_json_data(self):
        """
        Verify assemble_options fails when bad json data is provided.
        """
        # The input data must be a list of one or more serialize dicts.
        # Here we pass in variations that do not match what is expected.
        for tc in (
                True,
                None,
                [{'a': 'b'}],
            ):
            self.assertRaises(
                TypeError,
                cli.assemble_options, 
                [json.dumps(tc)])


test/test_kubeconfig.py
# Copyright (C) 2016-2017  Red Hat, Inc
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Library General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Library General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301  USA
"""
Test cases for the commctl.kubeconfig classes.
"""

import os
import mock

from . import TestCase, get_fixture_file_path
from commctl import kubeconfig


class TestKubeConfig(TestCase):
    """
    Tests for the KubeConfig class.
    """

    def test_kube_config__init(self):
        """
        Verify creating a new instance of KubeConfig works as it should.
        """
        file_path = get_fixture_file_path('test/kubeconfig.good')
        kc = kubeconfig.KubeConfig(file_path)
        kc.file_path = os.path.realpath(file_path)

    def test_kube_config__init_fails_with_a_missing_config(self):
        """
        Verify KubeConfig raises when the file path does not exist.
        """
        file_path = 'confi'
        self.assertRaises(
            IOError,
            kubeconfig.KubeConfig,
            file_path)

    def test_kube_config__init_fails_with_a_bad_config(self):
        """
        Verify KubeConfig raises when the configuration data is bad.
        """
        file_path = get_fixture_file_path('test/kubeconfig.bad')
        self.assertRaises(
            kubeconfig.KubeConfigInvalidFileError,
            kubeconfig.KubeConfig,
            file_path)

    def test_kube_config_current_context(self):
        """
        Verify KubeConfig.current_context returns the proper context.
        """
        expected = {
            'context': {
                'cluster': 'commissaire',
                'user': 'a'},
            'name': 'commissaire',
        }
        kc = kubeconfig.KubeConfig(
            get_fixture_file_path('test/kubeconfig.good'))
        self.assertEquals(expected, kc.current_context)

    def test_kube_config_current_user_with_token(self):
        """
        Verify KubeConfig.current_user returns the proper user data with token.
        """
        kc = kubeconfig.KubeConfig(
            get_fixture_file_path('test/kubeconfig.good'))
        self.assertEquals('token', kc.current_user.type)
        self.assertEquals('a', kc.current_user.token)
        self.assertEquals('a', kc.current_user.name)

    def test_kube_config_current_user_with_password(self):
        """
        Verify KubeConfig.current_user returns the proper user data with password.
        """
        kc = kubeconfig.KubeConfig(
            get_fixture_file_path('test/kubeconfig.password'))
        self.assertEquals('password', kc.current_user.type)
        self.assertEquals('a', kc.current_user.username)
        self.assertEquals('a', kc.current_user.password)
        self.assertEquals('a', kc.current_user.name)


src/commctl/helpers.py
# Copyright (C) 2016-2017  Red Hat, Inc
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Library General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Library General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301  USA
"""
Helper commands for use with the CLI but are not specific to the client.
"""


def do_passhash(args):
    """
    Uses bcrypt to hash a password.

    :param args: Parsed ArgumentParser args
    :type args: argparse.Namespace
    :returns: The hashed password
    :rtype: str
    """
    import bcrypt

    if args.password is not None:
        pw = args.password
    elif args.file is not None:
        pw = args.file.read()
    else:
        import getpass
        pw = getpass.getpass()
    salt = bcrypt.gensalt(log_rounds=args.rounds)
    return bcrypt.hashpw(pw, salt)


def do_user_data(args):
    """
    Generates a user-data file for cloud-init

    :param args: Parsed ArgumentParser args
    :type args: argparse.Namespace
    """
    import json
    import os.path
    import sys
    import tempfile

    from email.mime.multipart import MIMEMultipart
    from email.mime.text import MIMEText

    # Holds all thre required files to make our user-data
    # Keys are the mime-type, values are the paths to the files
    files = {
        'cloud-config': os.path.sep.join([
            os.path.dirname(os.path.realpath(
                __file__)), 'data', 'cloud-init.txt']),
        'x-commissaire-host': None,
        'part-handler': os.path.sep.join([os.path.dirname(
            os.path.realpath(__file__)), 'data', 'part-handler.py']),
    }

    if args.cloud_init is not None:
        files['cloud-config'] = args.cloud_init

    # Create and populate the configuration
    config_struct = {
        'endpoint': args.endpoint,
    }
    for key in (
        'username', 'password', 'cluster', 'remote_user',
            'ssh_key_path', 'authorized_keys_path'):
        value = getattr(args, key)
        if value is not None:
            config_struct[key] = value

    try:
        # Write the configuration out
        files['x-commissaire-host'] = tempfile.mktemp()
        with open(files['x-commissaire-host'], 'w') as f_obj:
            json.dump(config_struct, f_obj, indent=True)

        sub_messages = []
        for mtype, fname in files.items():
            with open(fname, 'r') as f_obj:
                sub_message = MIMEText(
                    f_obj.read(),
                    mtype,
                    sys.getdefaultencoding())
                sub_message.add_header(
                    'Content-Disposition',
                    'attachment; filename="{}"'.format(
                        os.path.basename(fname)))
                sub_messages.append(sub_message)
        # Pull it all together
        combined_message = MIMEMultipart()
        for msg in sub_messages:
            combined_message.attach(msg)

        # Write out to a file if output is set
        if args.outfile:
            # Write out the user-data file
            with open(args.outfile, 'w') as out:
                out.write(str(combined_message))
        else:
            sys.stdout.write(str(combined_message))
    finally:
        os.unlink(files['x-commissaire-host'])


src/commctl/__init__.py
# Copyright (C) 2016-2017  Red Hat, Inc
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Library General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Library General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301  USA
"""
Client CLI for commissaire.
"""


src/commctl/kubeconfig.py
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
Class for working with a kubeconfig file.
"""

import os
import yaml


class KubeConfigError(Exception):
    """
    Base error for all KubeConfig errors.
    """
    pass


class KubeConfigInvalidFileError(KubeConfigError):
    """
    Used when the kubeconfig file is invalid.
    """
    pass


class KubeConfigNoContextError(KubeConfigError):
    """
    Used when there is no context set.
    """
    pass


class KubeConfigUserTypeError(KubeConfigError):
    """
    Used when a user type can not be identified.
    """
    pass


class KubeConfigNoUserError(KubeConfigError):
    """
    Used when a user can not be found.
    """
    pass


class KubeUser:
    """
    Abstraction for a Kubernetes user.

    Each instance will have the following attributes:

    - name: The name of the user.
    - type: The type of the credentials: token, password, client-certificate
    - auth_headers: Additional authentication headers which should be used.
    """

    def __init__(self, spec):
        """
        Initializes a new KubeUser instance.

        :param spec: A kubeconfig user spec.
        :type spec: dict
        :raises: commctl.kubeconfig.KubeConfigUserTypeError
        """
        self.name = spec['name']

        if spec['user'].get('token'):
            self.token = spec['user']['token']
            self.auth_headers = [('Authorization', 'Bearer ' + self.token)]
            self.type = 'token'
        elif spec['user'].get('username') and spec['user'].get('password'):
            import base64
            import six
            self.username = spec['user']['username']
            self.password = spec['user']['password']
            user_pass = six.b('{}:{}'.format(self.username, self.password))
            basic_auth = base64.encodestring(user_pass)
            self.auth_headers = [(
                'Authorization', 'Basic ' + str(basic_auth))]
            self.type = 'password'
        elif (spec['user'].get('client-certificate') and
                spec['user'].get('client-key')):  # pragma: no cover
            self.client_certificate = os.path.realpath(
                spec['user']['client-certificate'])
            self.client_key = os.path.realpath(spec['user']['client-key'])
            # TODO
            self.auth_headers = ()
            self.type = 'client-certificate'
        else:
            raise KubeConfigUserTypeError(
                'Unknown user type: {0}'.format(spec))


class KubeConfig:
    """
    Abstraction for a kubeconfig file.
    """

    def __init__(self, file_path):
        """
        Initializes an instance of KubeConfig.

        :param file_path: Path to a kubeconfig file.
        :type file_path: str
        :raises: commctl.kubeconfig.KubeConfigInvalidFileError
        """
        self.file_path = os.path.realpath(file_path)
        with open(self.file_path, 'r') as f:
            try:
                self._kubeconfig = yaml.safe_load(f)
            except yaml.error.YAMLError as error:
                raise KubeConfigInvalidFileError(error)

    @property
    def current_context(self):
        """
        Returns the current context.

        :raises: commctl.kubeconfig.KubeConfigNoContextError
        """
        ctx_name = self._kubeconfig.get('current-context', None)
        for ctx in self._kubeconfig.get('contexts', []):
            if ctx_name == ctx['name']:
                return ctx
        raise KubeConfigNoContextError('No context has been set.')

    @property
    def current_user(self):
        """
        Returns the current user for the current context.

        :raises: commctl.kubeconfig.KubeConfigNoUserError
        """
        username = self.current_context.get('context', {}).get('user', None)
        for user_spec in self._kubeconfig.get('users', []):
            if username == user_spec['name']:
                return KubeUser(user_spec)
        raise KubeConfigNoUserError(
            'Can not find data for user {0}'.format(username))


src/commctl/hash_pass_script.py
# Copyright (C) 2016  Red Hat, Inc
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Library General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Library General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301  USA
"""
Utility to help generate bcrypt hashes.
"""

import argparse
import getpass
import sys

import bcrypt


def main():
    """
    Generates a bcrypt hash for a given password.
    """
    parser = argparse.ArgumentParser(
        description='bcrypt based password hasher.',
    )
    parser.add_argument('-r', '--rounds', type=int, default=12, help='')
    parser.add_argument(
        '--pwfile', type=argparse.FileType('r'), required=False,
        help='Reads password from a file instead of prompting.')
    args = parser.parse_args()

    password = None
    if args.pwfile:
        password = args.pwfile.readline().strip()
    else:
        password = getpass.getpass()

    hashed = bcrypt.hashpw(password, bcrypt.gensalt(log_rounds=args.rounds))
    sys.stdout.write('{0}\n'.format(hashed))


if __name__ == '__main__':  # pragma no cover
    main()


src/commctl/client_script.py
# Copyright (C) 2016-2017  Red Hat, Inc
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Library General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Library General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301  USA
"""
Client CLI for commissaire.
"""

from __future__ import print_function

import argparse

import commctl.cli


def main():
    """
    Main script entry point.
    """
    epilog = 'Example: commctl create deploy datacenter1 -u 7.2.2'

    parser = argparse.ArgumentParser(epilog=epilog)
    subparser = parser.add_subparsers(dest='command')

    cluster_parser = subparser.add_parser('cluster')
    commctl.cli.add_cluster_commands(cluster_parser)

    container_manager_parser = subparser.add_parser('container_manager')
    commctl.cli.add_container_manager_commands(container_manager_parser)

    host_parser = subparser.add_parser('host')
    commctl.cli.add_host_commands(host_parser)

    networks_parser = subparser.add_parser('network')
    commctl.cli.add_network_commands(networks_parser)

    # XXX passhash and user-data are more like helper scripts.  Keep them out
    #     of the shared API for now, and exclusive to commctl.
    subcmd_parser = subparser.add_parser('passhash')
    subcmd_parser.add_argument(
        '-p', '--password', help='Password to hash')
    subcmd_parser.add_argument(
        '-f', '--file', type=argparse.FileType('rb'),
        help='Password file to hash (or "-" for stdin)')
    subcmd_parser.add_argument(
        '-r', '--rounds', type=int, default=12, help='Number of rounds')

    user_data_parser = subparser.add_parser(
        'user-data',
        help='Generates a user-data file for use with cloud-init',
        epilog=(
            'Example: commctl user-data -p -c my_cluster -o cluster.userdata'))
    user_data_parser.add_argument(
        '-e', '--endpoint', required=True,
        help='Commissaire endpoint to use during bootstrapping')
    user_data_parser.add_argument(
        '-c', '--cluster', help='Name of the cluster for new hosts to join')
    user_data_parser.add_argument(
        '-u', '--username', help='Commissaire user to use when bootstrapping')
    user_data_parser.add_argument(
        '-p', '--password', action='store_true',
        help='Prompts for a Commissaire password to use when bootstrapping')
    user_data_parser.add_argument(
        '-r', '--remote-user',
        help='Remote user to provide to Commissaire for ssh access',
        default='root')
    user_data_parser.add_argument(
        '-s', '--ssh-key-path',
        help='Path to the private key of the remote user',
        default='/root/.ssh/id_rsa')
    user_data_parser.add_argument(
        '-a', '--authorized-keys-path',
        help='Path to the authorized_keys file of the remote user',
        default='/root/.ssh/authorized_keys')
    user_data_parser.add_argument(
        '-C', '--cloud-init', help='cloud-init.txt file to use')
    user_data_parser.add_argument(
        '-o', '--outfile', help='Output file. If omitted STDOUT is used')

    args = parser.parse_args()

    try:
        if args.command == 'passhash':
            from commctl.helpers import do_passhash
            print(do_passhash(args))
        elif args.command == 'user-data':
            from commctl.helpers import do_user_data
            from getpass import getpass
            if args.password:
                args.password = getpass()
            do_user_data(args)
        else:
            dispatcher = args._class()
            dispatcher.set_args(args)
            getattr(dispatcher, args.func)()
    except Exception as ex:
        parser.error(ex)


if __name__ == '__main__':  # pragma: no cover
    main()
